server/src/main/java/org/elasticsearch/action/support/broadcast/BroadcastRequest.java

```diff
@@ -20,56 +20,102 @@
+import java.util.ArrayList;
+import java.util.List;
 
-import org.elasticsearch.action.IndicesRequest;
+import org.elasticsearch.Version;
 import org.elasticsearch.action.support.IndicesOptions;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.transport.TransportRequest;
 
-public class BroadcastRequest<Request extends BroadcastRequest<Request>> extends TransportRequest implements IndicesRequest {
+import io.crate.metadata.IndexName;
+import io.crate.metadata.IndexParts;
+import io.crate.metadata.PartitionName;
+import io.crate.metadata.RelationName;
 
-    protected String[] indices;
-    private IndicesOptions indicesOptions = IndicesOptions.STRICT_EXPAND_OPEN_FORBID_CLOSED;
+public class BroadcastRequest extends TransportRequest {
 
-    public BroadcastRequest() {
+    protected final List<PartitionName> partitions;
+
+    protected BroadcastRequest(List<PartitionName> partitions) {
+        this.partitions = partitions;
     }
 
-    protected BroadcastRequest(String[] indices) {
-        this.indices = indices;
+    protected BroadcastRequest(PartitionName partition) {
+        this(List.of(partition));
     }
 
-    protected BroadcastRequest(String[] indices, IndicesOptions indicesOptions) {
-        this.indices = indices;
-        this.indicesOptions = indicesOptions;
+    public BroadcastRequest(StreamInput in) throws IOException {
+        super(in);
+        this.partitions = readPartitions(in);
     }
 
-    @Override
-    public String[] indices() {
-        return indices;
+    public final List<PartitionName> partitions() {
+        return partitions;
     }
 
     @Override
-    public IndicesOptions indicesOptions() {
-        return indicesOptions;
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        writePartitions(out, partitions);
+    }
+
+    private static final Version STREAMED_RELATIONSET_VERSION = Version.V_6_0_0;
+
+    private static List<PartitionName> readPartitions(StreamInput in) throws IOException {
+        if (in.getVersion().onOrAfter(STREAMED_RELATIONSET_VERSION)) {
+            return in.readList(PartitionName::new);
+        } else {
+            return readFromPre60(in);
+        }
     }
 
-    @SuppressWarnings("unchecked")
-    public final Request indicesOptions(IndicesOptions indicesOptions) {
-        this.indicesOptions = indicesOptions;
-        return (Request) this;
+    private static List<PartitionName> readFromPre60(StreamInput in) throws IOException {
+        List<PartitionName> relations = new ArrayList<>();
+        String[] indexes = in.readStringArray();
+        IndicesOptions.readIndicesOptions(in);
+        RelationName table = null;
+        List<String> partitionValues = new ArrayList<>();
+        for (String index : indexes) {
+            IndexParts indexParts = IndexName.decode(index);
+            if (table == null) {
```

Comment:
I don't understand this logic, what happens if `table` is already set and then we read and `index` which refers to a different partition of the same table?

Additionally, I think it would be nice to add a Streaming test for the `BroadRequest` .

Reply:
Yeah this was just wrong.  Have added a test and corrected the logic.